docs: OpenAI reponses endpoint (#9539)
langflow-ai/langflow: docs/docs/API-Reference/api-openai-responses.mdx

> ---

Author: Yes, great callout

title: OpenAI Responses API
slug: /api-openai-responses
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/Tabs';

> Langflow includes an OpenAI-compatible API endpoint that allows you to execute your flows using OpenAI-compatible request and response formats.

Author: improve this sentence and link to the OpenAI requests reference

> Langflow includes an OpenAI-compatible API endpoint that allows you to execute your flows using OpenAI-compatible request and response formats.

Reviewer: ```suggestion
Langflow includes an OpenAI-compatible API endpoint at `POST /api/v1/responses`.
This endpoint allows you to run your flows using OpenAI-compatible request and response formats.
```

This enables you to use existing OpenAI client libraries with minimal code changes - just change the `model` name to your `flow_id` instead of using OpenAI model names like "gpt-4" or "gpt-3.5-turbo".
Flow IDs can be found on the code snippets on the [**API access** pane](/concepts-publish#api-access) or in a flow's URL.

> ## Requirements

Reviewer: ```suggestion
## Prerequisites
```

Author: I feel like these are more "potential gotchas" than "prerequisites" so i called them Requirements

Your Langflow flow must contain a **ChatInput** component to be compatible with the OpenAI Responses API.
Flows without this component will return an error. The component types `ChatInput` and `Chat Input` are recognized as chat inputs.

- **Tools**: The `tools` parameter is not yet supported and will return an error if provided.
- **Flow Requirements**: Flows must contain a ChatInput component.
- **Model Names**: The `model` field must contain a valid flow ID or endpoint name.

> - **Authentication**: All requests require an API key passed in the `x-api-key` header:
> 
>     ```bash
>     curl -X POST \
>       "$LANGFLOW_SERVER_URL/api/v1/responses" \
>       -H "x-api-key: $LANGFLOW_API_KEY" \
>       -H "Content-Type: application/json" \
>       -d '{
>         "model": "your-flow-id",
>         "input": "Hello, how are you?"
>       }'
>     ```

Reviewer: ```suggestion
- **Authentication**: All requests require an API key passed in the `x-api-key` header.
For more information, see [API keys and authentication](/api-keys-and-authentication).
```

This avoids repeating the example multiple times in close proximity.

## Request

```
POST /api/v1/responses

> ```

Reviewer: I don't think you need this. I think you could just start with the example request and then explain the parameters (header & body).

Put all response stuff under a dedicated subsection.

## Headers

| Header | Required | Description | Example |
|--------|----------|-------------|---------|
| Content-Type | Yes | Specifies the JSON format | "application/json" |
| x-api-key | Yes | Your Langflow API key for authentication | "sk-..." |
| X-LANGFLOW-GLOBAL-VAR-* | No | Global variables for the flow | "X-LANGFLOW-GLOBAL-VAR-API_KEY: sk-..." |

## Example request

```curl
curl -X POST \
  "$LANGFLOW_SERVER_URL/api/v1/responses" \
  -H "x-api-key: $LANGFLOW_API_KEY" \
  -H "Content-Type: application/json" \
  -d '{
    "model": "your-flow-id",
    "input": "Hello, how are you?",
    "stream": false
  }'
```

<details>

> <summary>Response</summary>

Reviewer: ```suggestion
<summary>Result</summary>
```

To match the other pages, these should be labeled Result.
However, since you also describe the response contents, you could create a Response subsection that shows the example (without `<details>` and explains the content of the response.

{"id": "response-uuid", "object": "response.chunk", "created": 1703123457, "model": "your-flow-id", "delta": {"content": " upon"}}
{"id": "response-uuid", "object": "response.chunk", "created": 1703123458, "model": "your-flow-id", "delta": {"content": " a"}}
{"id": "response-uuid", "object": "response.chunk", "created": 1703123459, "model": "your-flow-id", "delta": {"content": " time"}}
{"id": "response-uuid", "object": "response.chunk", "created": 1703123460, "model": "your-flow-id", "delta": {"content": "..."}}
{"id": "response-uuid", "object": "response.chunk", "created": 1703123461, "model": "your-flow-id", "delta": {}, "status": "completed"}
```

</details>
### Streaming response body

| Field | Type | Description |
|-------|------|-------------|
| id | string | Unique response identifier |
| object | string | Always "response.chunk" |
| created | int | Unix timestamp of chunk creation |
| model | string | The flow ID that was executed |
| delta | dict | The new content chunk |
| status | string | Response status: "completed", "in_progress", or "failed" (optional) |

## Pass global variables to your flows in headers

> 

Reviewer: Does this mean you are setting values for global variables that are named in the flow? It will override any value set in Langflow itself?

I just want to understand clearly: You can't arbitrarily declare a global variable that isn't declared in the flow or related to flow execution?

Author: Explained below

This is useful for passing API keys, user IDs, or any other configuration that might change between requests.

The `/responses` endpoint accepts global variables as custom HTTP headers with the format `X-LANGFLOW-GLOBAL-VAR-{VARIABLE_NAME}`.
Variable names are automatically converted to uppercase.

This example passes three global variables as headers.
In your flow, these variables will be available as `API_KEY` (from `X-LANGFLOW-GLOBAL-VAR-API_KEY`), `USER_ID` (from `X-LANGFLOW-GLOBAL-VAR-USER_ID`), and `ENVIRONMENT` (from `X-LANGFLOW-GLOBAL-VAR-ENVIRONMENT`).

```bash
curl -X POST \
  "$LANGFLOW_SERVER_URL/api/v1/responses" \
  -H "x-api-key: $LANGFLOW_API_KEY" \
  -H "Content-Type: application/json" \
  -H "X-LANGFLOW-GLOBAL-VAR-API_KEY: sk-..." \
  -H "X-LANGFLOW-GLOBAL-VAR-USER_ID: user123" \
  -H "X-LANGFLOW-GLOBAL-VAR-ENVIRONMENT: production" \
  -d '{
    "model": "your-flow-id",
    "input": "Hello"
  }'
```

## Retrieve tool call results

When your flow includes tools or function calls, you can get detailed information about their execution by including `"tool_call.results"` in the `include` parameter.

> #### Basic Tool Call Response (Default)

Reviewer: ```suggestion
```

I don't think this subheading is necessary unless you also add one for the other option.

Without the `include` parameter, tool calls return basic function call information:

```json
{
  "id": "fc_1",
  "type": "function_call",
  "status": "completed",
  "name": "get_weather",
  "arguments": "{\"location\": \"New York\"}"
}
```

With `include: ["tool_call.results"]`, you get comprehensive information about tool execution:

```json
{
  "id": "get_weather_fc_1",
  "queries": ["New York"],
  "status": "completed",
  "tool_name": "get_weather",
  "type": "tool_call",
  "results": {
    "temperature": 72,
    "condition": "sunny",
    "humidity": 45,
    "wind_speed": 8
  }
}
```

> ## Conversation continuity using `previous_response_id`

Reviewer: ```suggestion
## Continue conversations with response and session IDs
```

This heading avoids the code font and accounts for both options (response ID and session ID).

When you make a request, the API returns a response with an `id` field. You can use this `id` as the `previous_response_id` in your next request to continue the conversation from where it left off.

First Message:
```bash
curl -X POST \
  "$LANGFLOW_SERVER_URL/api/v1/responses" \
  -H "x-api-key: $LANGFLOW_API_KEY" \
  -H "Content-Type: application/json" \
  -d '{
    "model": "your-flow-id",
    "input": "Hello, my name is Alice"
  }'
```

<details>
<summary>Response</summary>

```json
{
  "id": "response-abc123",
  "object": "response",
  "status": "completed",
  "model": "your-flow-id",
  "output": [...],
  "previous_response_id": null
}
```

</details>

Follow-up message:
```bash
curl -X POST \
  "$LANGFLOW_SERVER_URL/api/v1/responses" \
  -H "x-api-key: $LANGFLOW_API_KEY" \
  -H "Content-Type: application/json" \
  -d '{
    "model": "your-flow-id",
    "input": "What's my name?",
    "previous_response_id": "response-abc123"
  }'
```

<details>
<summary>Response</summary>

```json
{
  "id": "response-def456",
  "object": "response",
  "status": "completed",
  "model": "your-flow-id",
  "output": [...],
  "previous_response_id": "response-abc123"
}
```

</details>

> You can also use your own session IDs instead of the response ID:

Reviewer: Mention how session ID relates to chat memory.

The default values are documented here for completeness and to show the full response structure.

Fields set dynamically by Langflow:

| Field | Type | Description |
|-------|------|-------------|
| id | string | Unique response identifier. |
| created_at | int | Unix timestamp of response creation. |
| model | string | The flow ID that was executed. |
| output | list[dict] | Array of output items (messages, tool calls, etc.). |
| previous_response_id | string | ID of previous response if continuing conversation. |

<details closed>
<summary>Fields with OpenAI-compatible default values</summary>

| Field | Type | Default Value | Description |
|-------|------|---------------|-------------|
| object | string | "response" | Always "response". |
| status | string | "completed" | Response status: "completed", "in_progress", or "failed". |
| error | dict | null | Error details (if any). |
| incomplete_details | dict | null | Incomplete response details (if any). |
| instructions | string | null | Response instructions (if any). |
| max_output_tokens | int | null | Maximum output tokens (if any). |
| parallel_tool_calls | boolean | true | Whether parallel tool calls are enabled. |

> | reasoning | dict | `{"effort": null, "summary": null}` | Reasoning information with effort and summary. |

Author: Fixed, thank you

---
title: OpenAI Responses API
slug: /api-openai-responses
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/Tabs';

Langflow includes an OpenAI-compatible API endpoint at `POST /api/v1/responses`.

> This endpoint allows you to run your flows using OpenAI-compatible request and response formats.

Reviewer 2: you're overloading the term responses here. the responses api vs request / response format.
